Automate docs with eslint-doc-generator

bmish opened the pull request:
I built this CLI tool [eslint-doc-generator](https://github.com/bmish/eslint-doc-generator) for automating the generation of the README rules list table and rule doc title/notices for ESLint plugins. It follows common documentation conventions from this and other top ESLint plugins and will help us standardize documentation across ESLint plugins (and generally improve the usability of custom rules through better documentation and streamline the process of adding new rules). It has 100% test coverage and is already used in many popular ESLint plugins.

dangreenisrael commented:
Also, if you felt like upgrading the base image in a separate PR that would be more than welcome as well :D

bmish commented on package.json:
```diff
@@ -38,15 +40,16 @@
     "eslint-plugin-prettier": "^4.0.0",
     "jest": "^27.3.1",
     "prettier": "2.4.1",
-    "typescript": "^4.4.4",
```
The dependencies were automatically sorted when I ran `npm i --save-dev eslint-doc-generator`.

bmish commented:
@dangreenisrael 

> If you want to do a cleanup to remove old rules, could you do that in a separate PR?

Removing docs for deleted rules in https://github.com/dangreenisrael/eslint-plugin-jest-formatting/pull/118.

> Also, if you felt like upgrading the base image in a separate PR that would be more than welcome as well 😄

What do you mean exactly? 

Regarding CI, eslint-doc-generator requires Node >= 14, causing CI to fail when installing deps and running linting on node 12. I split out the lint job to only run under Node 18 and switched to [`yarn install --ignore-engines` ](https://classic.yarnpkg.com/lang/en/docs/cli/install/#toc-yarn-install-ignore-engines) to install dependencies.

bmish commented:
Can someone approve CI to run?

bmish commented:
Can someone approve CI to run? This is still ready to go in. @dangreenisrael @twelve17 @hockeybuggy

benkimpel commented:
@bmish I've approved the CI run, _however_, I'd like to leave it up to @dangreenisrael to do the merge on this one since this PR could impact how he manages this repo (generating the docs, for example.)

benkimpel commented:
@bmish I also rebased this onto latest after merging your other PR.

dangreenisrael commented:
Are we still looking to land this?

bmish commented:
@dangreenisrael yes still ready to go.
